at-angular.ts

```diff
@@ -107,8 +151,36 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
+
+        return (targetClass: any): void => {
+
+            angular
+                .module(moduleName)
+                .config([
+                    '$provide',
+                    function($provide: angular.auto.IProvideService): void {
+
+                    $provide.decorator(targetProvider, [
+                       '$delegate',
+                        ($delegate: any): void => {
+
+                            $delegate[targetProvider] = $delegate;
+
+                            angular.extend($delegate, new targetClass());
```

Comment:
But in this case, you only have a $delegate as first parameter. Isn't the need. I need the injection on the targetClass contructor (like a service) to get other services.

$delegate = angular.extend(
    $delegate, 
    targetClass.prototype, 
    new targetClass($delegate)
);
I don't understand why you set targetClass.prototype and new targetClass($delegate) as parameters... 
new targetClass($delegate) return an object/instance with method and prototype return an object with method. Isn't clear for me...